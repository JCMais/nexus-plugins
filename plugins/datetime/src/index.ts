import {
  core,
  plugin,
  dynamicOutputMethod,
  objectType,
  stringArg,
  interfaceType,
  nonNull,
} from 'nexus'

import * as moment from 'moment-timezone'

export type DateTimePluginConfig = {
  /**
   * The name of the dateTime field, defaults to "dateTime"
   */
  dateTimeFieldName?: string

  nexusSchemaImportId?: string
  dateTimePluginImportId?: string
}

export type DateTimePluginFieldConfig<
  TypeName extends string = any,
  FieldName extends string = any
> = {
  /**
   * This defaults to the fieldName used when calling dateTime(fieldName)
   */
  dateTimeISOField?: string
} & Exclude<NexusGenPluginFieldConfig<TypeName, FieldName>, 'resolve' | 'type'>

export function dateTimePlugin(pluginConfig: DateTimePluginConfig = {}) {
  const {
    dateTimeFieldName = 'dateTime',
    dateTimePluginImportId = '@jcm/nexus-plugin-datetime',
  } = pluginConfig

  return plugin({
    name: 'DateTime',
    description: 'add t.dateTime(field) to the schema builder',
    // we want to add a extension
    fieldDefTypes: [
      core.printedGenTypingImport({
        module: dateTimePluginImportId,
        bindings: ['DateTimePluginFieldConfig'],
      }),
    ],
    onInstall(builder) {
      builder.addType(
        dynamicOutputMethod({
          name: dateTimeFieldName,
          typeDefinition: `<FieldName extends string>(
            fieldName: FieldName, 
            config: DateTimePluginFieldConfig<TypeName, FieldName>
          ): void`,
          factory({ typeName: _parentTypeName, typeDef: t, args: factoryArgs }) {
            const [fieldName, fieldConfig = {}] = factoryArgs
            const { dateTimeISOField = fieldName, ...remainingFieldConfig } = fieldConfig

            if (!builder.hasType('DateTimeFieldInterface')) {
              builder.addType(
                interfaceType({
                  name: 'DateTimeFieldInterface',
                  description: 'A object that represents an ISO datetime',
                  resolveType(obj) {
                    if (obj.iso) return 'DateTimeField'
                    return null
                  },
                  definition(t2) {
                    t2.nonNull.string('iso')
                    t2.nonNull.string('formatted', {
                      args: {
                        format: nonNull(stringArg()),
                        timezone: stringArg({
                          description: 'Timezone to format the ISO date, defaults to UTC',
                        }),
                      },
                      resolve: (root, args) => {
                        return args.timezone
                          ? moment.tz(root.iso, args.timzone).format(args.format)
                          : moment.utc(root.iso).format(args.format)
                      },
                    })
                    t2.nonNull.boolean('isAfter', {
                      args: {
                        iso: stringArg({
                          description: 'Defaults to the current time if not provided',
                        }),
                      },
                      resolve: (root, { iso }) => moment(root.iso).isAfter(iso || new Date()),
                    })
                    t2.nonNull.boolean('isBefore', {
                      args: {
                        iso: stringArg({
                          description: 'Defaults to the current time if not provided',
                        }),
                      },
                      resolve: (root, { iso }) => moment(root.iso).isBefore(iso || new Date()),
                    })
                    t2.nonNull.boolean('isSameOrAfter', {
                      args: {
                        iso: stringArg({
                          description: 'Defaults to the current time if not provided',
                        }),
                      },
                      resolve: (root, { iso }) => moment(root.iso).isSameOrAfter(iso || new Date()),
                    })
                    t2.nonNull.boolean('isSameOrBefore', {
                      args: {
                        iso: stringArg({
                          description: 'Defaults to the current time if not provided',
                        }),
                      },
                      resolve: (root, { iso }) =>
                        moment(root.iso).isSameOrBefore(iso || new Date()),
                    })
                    t2.nonNull.boolean('isSame', {
                      args: {
                        iso: stringArg({
                          description: 'Defaults to the current time if not provided',
                        }),
                      },
                      resolve: (root, { iso }) => moment(root.iso).isSame(iso || new Date()),
                    })
                    t2.nonNull.boolean('isBetween', {
                      args: {
                        isoStart: stringArg({
                          description: 'Defaults to the start of the current day if not provided',
                        }),
                        isoEnd: stringArg({
                          description: 'Defaults to the end of the current day if not provided',
                        }),
                      },
                      resolve: (root, { isoStart, isoEnd }) =>
                        moment(root.iso).isBetween(
                          isoStart || moment().startOf('day'),
                          isoEnd || moment().endOf('day'),
                        ),
                    })
                  },
                }),
              )
            }

            if (!builder.hasType('DateTimeField')) {
              builder.addType(
                objectType({
                  name: 'DateTimeField',
                  description: 'Represents an ISO datetime that can be formatted in other formats',
                  definition: (t2) => {
                    t2.implements('DateTimeFieldInterface')
                  },
                }),
              )
            }

            t.field(fieldName, {
              ...remainingFieldConfig,
              type: 'DateTimeField',
              resolve: (root) =>
                root[dateTimeISOField] && {
                  iso: moment(root[dateTimeISOField]).toISOString(),
                },
            })
          },
        }),
      )
    },
  })
}
