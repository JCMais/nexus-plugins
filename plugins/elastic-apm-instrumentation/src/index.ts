import { plugin, core } from '@nexus/schema'
import {
  isLeafType,
  isWrappingType,
  ResponsePath,
  GraphQLOutputType,
  GraphQLWrappingType,
} from 'graphql'
import type * as Agent from 'elastic-apm-node'

// Most of this code is inspired on the authorize plugin from @nexus/schema

const fieldDefTypes = core.printedGenTyping({
  optional: true,
  name: 'shouldCreateApmSpan',
  description: `
    This can be used to override the elastic apm instrumentation settings and force
    a span to [not] be created for this field.
  `,
  type: 'boolean',
})

export interface ElasticApmInstrumentationPluginConfig {
  apmAgent: typeof Agent

  /**
   * Should create spans by default for all fields that pass the ignore checks?
   *
   * Defaults to true
   */
  shouldCreateApmSpanByDefault?: boolean

  /**
   * Should leaf fields be ignored?
   *
   * Defaults to true
   */
  shouldIgnoreLeafFieldsByDefault?: boolean

  /**
   * Pass an array of fields to be ignored from span creation, each item can be a string or a RegExp
   * If a string, it's also possible to use wildcards, like: `PageInfo.*`
   */
  ignoredFields?: Array<string | RegExp>

  /**
   * Pass an array of types to be ignored from span creation, each item can be a string or a RegExp
   */
  ignoredTypes?: Array<string | RegExp>
}

const isValidMatch = (strA: string, strB: string) => strA === strB || strA === '*'

const isPatternMatchingField = (pattern: string, field: string) => {
  if (pattern === '*') return true

  const piecesPatternPieces = pattern.split('.')
  const fieldPatternPieces = field.split('.')

  if (piecesPatternPieces.length !== 2 || fieldPatternPieces.length !== 2) {
    throw new Error('Pattern/Field path must have exactly 2 parts, ParentType.childType')
  }

  return (
    isValidMatch(piecesPatternPieces[0], fieldPatternPieces[0]) &&
    isValidMatch(piecesPatternPieces[1], fieldPatternPieces[1])
  )
}

const getFirstNonWrappingType = (
  parent: GraphQLOutputType,
): Exclude<GraphQLOutputType, GraphQLWrappingType> => {
  if (isWrappingType(parent)) return getFirstNonWrappingType(parent.ofType)

  return parent
}

const getGraphQLFieldPath = (path?: ResponsePath): string =>
  (path
    ? [
        getGraphQLFieldPath(path.prev),
        // @ts-expect-error
        path.prev ? path.key : `${path.typename}.${path.key}`,
      ].filter((v) => !!v)
    : []
  ).join('.')

export const elasticApmInstrumentationPlugin = (
  pluginConfig: ElasticApmInstrumentationPluginConfig,
) => {
  const {
    apmAgent,
    shouldCreateApmSpanByDefault = true,
    shouldIgnoreLeafFieldsByDefault = true,
    ignoredFields = [],
    ignoredTypes = [],
  } = pluginConfig

  return plugin({
    name: 'ElasticApmReporter',
    description: 'Creates separated spans for each resolved field',
    fieldDefTypes: fieldDefTypes,
    onCreateFieldResolver(config) {
      const firstNonWrappingType = getFirstNonWrappingType(config.fieldConfig.type)
      const isLeaf = isLeafType(firstNonWrappingType)

      const parentTypeName = config.parentTypeConfig.name
      const fieldName = config.fieldConfig.name
      const qualifiedName = `${parentTypeName}.${fieldName}`

      const isIgnoredField = ignoredFields.some((ignoredField) =>
        ignoredField instanceof RegExp
          ? ignoredField.test(qualifiedName)
          : isPatternMatchingField(ignoredField, qualifiedName),
      )

      const isIgnoredType = ignoredTypes.some((type) =>
        type instanceof RegExp
          ? type.test(firstNonWrappingType.name)
          : type === firstNonWrappingType.name,
      )

      const shouldCreateApmSpan =
        config.fieldConfig.extensions?.nexus?.config?.shouldCreateApmSpan ??
        (shouldCreateApmSpanByDefault &&
          (!shouldIgnoreLeafFieldsByDefault || !isLeaf) &&
          !isIgnoredField &&
          !isIgnoredType)

      // console.log({
      //   qualifiedName,
      //   shouldCreateApmSpan,
      //   firstNonWrappingType,
      //   isLeaf,
      //   isIgnoredField,
      //   isIgnoredType,
      //   type: config.fieldConfig.type,
      // })

      if (!shouldCreateApmSpan) {
        return
      }

      return function (root, args, ctx, info, next) {
        const isApmAgentStartedAndHasTransaction =
          apmAgent.isStarted() && apmAgent.currentTransaction

        if (!isApmAgentStartedAndHasTransaction) {
          return next(root, args, ctx, info)
        }

        const fieldPath = getGraphQLFieldPath(info.path)
        const span = apmAgent.startSpan(
          `GraphQL resolver: ${fieldPath}`,
          'db',
          'graphql',
          'resolve',
        )

        let toComplete
        try {
          toComplete = next(root, args, ctx, info)
        } catch (e) {
          toComplete = Promise.reject(e)
        }
        return plugin.completeValue(
          toComplete,
          (val) => {
            span?.end()
            return val
          },
          (error) => {
            span?.end()
            throw error
          },
        )
      }
    },
  })
}
