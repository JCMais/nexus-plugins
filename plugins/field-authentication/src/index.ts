import { plugin, core } from '@nexus/schema'
import { GraphQLResolveInfo } from 'graphql'
import { MaybePromise } from '@nexus/schema/dist/core'

// Most of this code is inspired on the authorize plugin from @nexus/schema

const fieldAuthenticationPluginResolverImport = core.printedGenTypingImport({
  module: '@jcm/nexus-plugin-field-authentication',
  bindings: ['FieldAuthenticationResolver'],
})

const fieldDefTypes = core.printedGenTyping({
  optional: true,
  name: 'authentication',
  description: `
    Authentication for an individual field. Returning "true"
    or "Promise<true>" means the field can be accessed only if
    the request is authenticated. Returning "false" or "Promise<false>"
    means the field can be acessed only if the request is *NOT* authenticated.
    It's also possible to return a tuple "[boolean, Error | ResolveValue]"
    where the first element specifies if the request should be authenticated
    or not and the second specifies the behavior:
    If the second element is an error, this error will be returned for the field, even if
      throwErrorOnFailedAuthentication is false.
    If the second element is anything else, this value will be resolved if the request
     fails the authentication check.
  `,
  type: 'FieldAuthenticationResolver<TypeName, FieldName>',
  imports: [fieldAuthenticationPluginResolverImport],
})

export type FieldAuthenticationResolverReturnValue =
  | boolean
  | Error
  | [boolean, Error]
  | [boolean, core.ResultValue<any, any>]

export type FieldAuthenticationResolver<TypeName extends string, FieldName extends string> = (
  root: core.RootValue<TypeName>,
  args: core.ArgsValue<TypeName, FieldName>,
  context: core.GetGen<'context'>,
  info: GraphQLResolveInfo,
) => core.MaybePromise<FieldAuthenticationResolverReturnValue>

export interface FieldAuthenticationPluginErrorConfig {
  error: Error
  root: any
  args: any
  ctx: core.GetGen<'context'>
  info: GraphQLResolveInfo
}

export const defaultFormatError = ({ error }: FieldAuthenticationPluginErrorConfig): Error => error

export type FieldAuthenticationPluginConfigIsLogged<
  TypeName extends string,
  FieldName extends string
> = (
  root: core.RootValue<TypeName>,
  args: core.ArgsValue<TypeName, FieldName>,
  context: core.GetGen<'context'>,
  info: GraphQLResolveInfo,
) => boolean | MaybePromise<boolean>

export interface FieldAuthenticationPluginConfig {
  /**
   * This will be called with the error that will be returned when resolving a field
   *  that did not pass the authentication check when throwErrorOnFailedAuthenticationByDefault
   *  was set to true or when the authentication had a tuple [boolean, Error]
   * By default this is the function ({error: Error}) => error
   */
  formatError?: (authConfig: FieldAuthenticationPluginErrorConfig) => Error
  /**
   * If this is true and the authentication field does not include
   *  a default resolve value, an error will be thrown on failed authentication
   */
  throwErrorOnFailedAuthenticationByDefault?: boolean
  /**
   * If throwErrorOnFailedAuthenticationByDefault is true,
   */
  defaultErrorMessage?: string
  /**
   * If throwErrorOnFailedAuthenticationByDefault is false, a failed authentication check will return
   *  this value as the resolved value of the field
   */
  defaultResolveValue?: core.ResultValue<any, any>
  /**
   * By default this checks if ctx.state.user exists
   */
  isLogged?: FieldAuthenticationPluginConfigIsLogged<any, any>
}

// This is much more complex than it could be
//  but it was done that way so that it handle all scenarios possible that we might ever encounter.

export function fieldAuthenticationPlugin(pluginConfig: FieldAuthenticationPluginConfig = {}) {
  const {
    formatError = defaultFormatError,
    throwErrorOnFailedAuthenticationByDefault = false,
    defaultErrorMessage = 'Not Authenticated',
    defaultResolveValue = null,
    isLogged = (_root, _args, ctx, _info) => !!ctx?.state?.user,
  } = pluginConfig

  const ensureError = (
    root: any,
    args: any,
    ctx: core.GetGen<'context'>,
    info: GraphQLResolveInfo,
  ) => (error: Error) => {
    const finalErr = formatError({ error, root, args, ctx, info })
    if (finalErr instanceof Error) {
      throw finalErr
    }
    ;(ctx.logger || console).error(
      `Non-Error value ${JSON.stringify(
        finalErr,
      )} returned from custom formatError in field authentication plugin`,
    )
    throw new Error(defaultErrorMessage)
  }

  return plugin({
    name: 'FieldAuthentication',
    // we want to add a extension
    description: 'Makes sure request is authenticated before calling next resolvers in the chain',
    fieldDefTypes: fieldDefTypes,
    onCreateFieldResolver(config) {
      const authentication = config.fieldConfig.extensions?.nexus?.config?.authentication

      if (typeof authentication !== 'undefined') {
        // The authentication wrapping resolver.
        return function (root, args, ctx, info, next) {
          const {
            fieldName,
            parentType: { name: parentTypeName },
          } = info

          const processAuthenticationResult = (
            isUserLogged: boolean,
            result: FieldAuthenticationResolverReturnValue,
          ) => {
            const finalFormatError = ensureError(root, args, ctx, info)

            if (
              typeof result === 'boolean' ||
              (Array.isArray(result) && result.length === 2 && typeof result[0] === 'boolean')
            ) {
              const canProceed =
                isUserLogged === result || (Array.isArray(result) && isUserLogged === result[0])

              // all branches here must return finalFormatError
              if (!canProceed) {
                if (Array.isArray(result)) {
                  if (result[1] instanceof Error) {
                    return finalFormatError(result[1])
                  }
                  return result[1]
                } else if (throwErrorOnFailedAuthenticationByDefault) {
                  return finalFormatError(new Error(defaultErrorMessage))
                }
                return defaultResolveValue
              }

              return next(root, args, ctx, info)
            } else {
              return finalFormatError(
                new Error(
                  `Field authentication for ${parentTypeName}.${fieldName} expected a boolean or [boolean, resolvedValue] tuple, saw ${JSON.stringify(
                    result,
                  )} instead`,
                ),
              )
            }
          }

          let toCompleteIsLogged
          try {
            toCompleteIsLogged = isLogged(root, args, ctx, info)
          } catch (e) {
            toCompleteIsLogged = Promise.reject(e)
          }
          return plugin.completeValue(
            toCompleteIsLogged,
            (isUserLogged) => {
              if (typeof authentication !== 'function') {
                return processAuthenticationResult(isUserLogged, authentication)
              }

              let toComplete
              try {
                toComplete = authentication(root, args, ctx, info)
              } catch (e) {
                toComplete = Promise.reject(e)
              }
              return plugin.completeValue(
                toComplete,
                processAuthenticationResult.bind(undefined, isUserLogged),
                // @ts-expect-error
                (err) => {
                  ensureError(root, args, ctx, info)(err as Error)
                },
              )
            },
            // @ts-expect-error
            (err) => {
              ensureError(root, args, ctx, info)(err)
            },
          )
        }
      }
    },
  })
}
