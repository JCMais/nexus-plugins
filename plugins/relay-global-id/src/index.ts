import { GraphQLResolveInfo } from 'graphql'
import { plugin, dynamicOutputMethod, core } from '@nexus/schema'
import { printedGenTypingImport } from '@nexus/schema/dist/utils'

import { toGlobalId } from 'graphql-relay'

export interface RelayGlobalIdPluginConfig {
  nexusFieldName?: string
  nexusSchemaImportId?: string
  relayGlobalIdPluginImportId?: string
}

export type RelayGlobalIdNexusFieldConfig<
  TypeName extends string = any,
  FieldName extends string = any
> = {
  /**
   * If no resolve is supplied, the default value to be used as global ID will be root[field]
   *
   * This defaults to the fieldName used when calling relayGlobalId(fieldName)
   *
   * If a resolve is passed, this is ignored
   */
  field?: string

  /**
   * Defaults to true - Adds a new ID! field called `raw${uppercase(fieldName)}`, where `fieldName` is `t.relayGlobalId(fieldName)`,
   *  with the original value of the field.
   *
   * It's also possible to pass a string with a different name for the field
   */
  shouldAddRawId?: boolean | string
  /**
   * Defaults to the parent type of the current field.
   */
  typeName?: string
  resolve?: (
    root: core.RootValue<TypeName>,
    args: core.ArgsValue<TypeName, FieldName>,
    ctx: core.GetGen<'context'>,
    info: GraphQLResolveInfo,
  ) =>
    | core.MaybePromise<core.ResultValue<TypeName, FieldName>>
    | core.MaybePromiseDeep<core.ResultValue<TypeName, FieldName>>
} & NexusGenPluginFieldConfig<TypeName, FieldName>

export function relayGlobalIdPlugin(pluginConfig: RelayGlobalIdPluginConfig) {
  const {
    nexusFieldName = 'relayGlobalId',
    nexusSchemaImportId = '@jcm/relay-global-id',
    relayGlobalIdPluginImportId = '@jcm/nexus-plugin-relay-global-id',
  } = pluginConfig

  return plugin({
    name: 'RelayGlobalId',
    description: 'add t.relayGlobalId(field) to the schema builder',
    fieldDefTypes: [
      printedGenTypingImport({
        module: nexusSchemaImportId,
        bindings: ['core'],
      }),
      printedGenTypingImport({
        module: relayGlobalIdPluginImportId,
        bindings: ['RelayGlobalIdNexusFieldConfig'],
      }),
    ],
    // we want to add a extension
    onInstall(builder) {
      builder.addType(
        dynamicOutputMethod({
          name: nexusFieldName,
          typeDefinition: `<FieldName extends string>(
            fieldName: FieldName, 
            config: RelayGlobalIdNexusFieldConfig<TypeName, FieldName>
          ): void`,
          factory({ typeName: parentTypeName, typeDef: t, args: factoryArgs }) {
            const [fieldName, fieldConfig = {}] = factoryArgs as [
              string,
              RelayGlobalIdNexusFieldConfig,
            ]

            const {
              field = fieldName,
              shouldAddRawId = true,
              typeName = parentTypeName,
              resolve: resolveFn,
            } = fieldConfig

            t.id(fieldName, {
              async resolve(root, args, ctx, info) {
                return toGlobalId(
                  typeName,
                  resolveFn ? await resolveFn(root, args, ctx, info) : root[field],
                )
              },
            })

            if (shouldAddRawId) {
              t.id(
                typeof shouldAddRawId === 'string'
                  ? shouldAddRawId
                  : `raw${fieldName.charAt(0).toUpperCase()}${fieldName.slice(1)}`,
                {
                  resolve: resolveFn ? resolveFn : (root) => root[field],
                },
              )
            }
          },
        }),
      )

      // TODO: Deprecate this syntax
      return { types: [] }
    },
  })
}
