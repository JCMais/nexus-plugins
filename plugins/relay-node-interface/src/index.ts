import { plugin, core, interfaceType, queryField, idArg } from '@nexus/schema'
import { fromGlobalId } from 'graphql-relay'
import { GraphQLResolveInfo } from 'graphql'

export type RelayNodeInterfacePluginConfig = {
  idFetcher: (
    val: { id: string; type: core.AllOutputTypes },
    ctx: core.GetGen<'context'>,
    info: GraphQLResolveInfo,
  ) => core.ResultValue<any, any>
  resolveType: (object: core.ResultValue<any, any>) => core.AllOutputTypes
  /**
   * Used to parse the ID before calling idFetcher - By default this calls fromGlobalId from graphql-relay
   */
  idParser?: (id: string) => any
}

// Pretty much based on nodeDefinitions function from
//  relay: https://github.com/graphql/graphql-relay-js/blob/8f4ed1ad35805ef233ad9fd1af33abb9c0cad35a/src/node/node.js
export function relayNodeInterfacePlugin(pluginConfig: RelayNodeInterfacePluginConfig) {
  const { idFetcher, resolveType, idParser = fromGlobalId } = pluginConfig

  if (!idFetcher) {
    throw new Error('idFetcher option is required for relayNodeInterfacePlugin')
  }

  if (typeof idFetcher !== 'function') {
    throw new Error(
      'idFetcher option must be a function with signature: async ({ id, type }, ctx, info) => T where T is the resolved value for the ID',
    )
  }

  if (!resolveType) {
    throw new Error('resolveType option is required for relayNodeInterfacePlugin')
  }

  if (typeof resolveType !== 'function') {
    throw new Error(
      'resolveType option must be an function with signature async (value) => T where T is the type name or object',
    )
  }

  return plugin({
    name: 'RelayNodeInterface',
    description: 'Creates the Relay Node interface and add node/nodes fields to the Query type',
    onInstall(builder) {
      if (!builder.hasType('Node')) {
        // node interface
        builder.addType(
          interfaceType({
            name: 'Node',
            description: 'An object with a global ID',
            definition: (t) => {
              t.id('id', {
                description: 'The global ID of the object.',
              })

              // overwrite the resolve type with the client function
              // resolveType is just one way to map objs -> to the graphql type they represent
              //  we could also have simply relied on isTypeOf on the GraphQLObjectType themselves
              //  but as relay uses this by default, let's keep using it
              t.resolveType(resolveType)
            },
          }),
        )

        // node field
        builder.addType(
          queryField((t) => {
            t.field('node', {
              type: 'Node',
              nullable: true,
              args: {
                id: idArg({
                  required: true,
                  description: 'The global ID of an object',
                }),
              },
              description: 'Fetches an object given its global ID',
              resolve: (_obj, { id }, context, info) => idFetcher(fromGlobalId(id), context, info),
            })
          }),
        )

        // nodes field
        builder.addType(
          queryField((t) => {
            t.field('nodes', {
              type: 'Node',
              list: [false],
              args: {
                ids: idArg({
                  list: [true],
                  required: true,
                  description: 'The global IDs of objects',
                }),
              },
              description: 'Fetches objects given their global IDs',
              resolve: (_obj, { ids }, context, info) =>
                Promise.all(
                  // @ts-expect-error
                  ids.map((id) => Promise.resolve(idFetcher(idParser(id), context, info))),
                ),
            })
          }),
        )
      }

      // TODO: Deprecate this syntax
      return { types: [] }
    },
  })
}
