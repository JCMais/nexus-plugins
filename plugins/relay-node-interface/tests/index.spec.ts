import { execute, parse, printType } from 'graphql'
import { toGlobalId } from 'graphql-relay'
import { makeSchema, objectType } from '@nexus/schema'

import { RelayNodeInterfacePluginConfig, relayNodeInterfacePlugin } from '../src'

const users = [
  {
    id: '8a291df8-bbcf-11ea-9db0-ff6b23713451',
    name: 'user-a',
  },
  {
    id: '710c15fa-c2af-11ea-86f1-7f51ff98b69c',
    name: 'user-b',
  },
  {
    id: '710c16cc-c2af-11ea-86f1-479adc4ed46f',
    name: 'user-c',
  },
]

const posts = [
  { id: '710c173a-c2af-11ea-86f1-83226be355fb', title: 'post-a' },
  { id: 'a7a5c06e-c2d9-11ea-86f1-bfb838f3b7c6', title: 'post-b' },
]

const comments = [
  {
    id: 'abc',
    text: 'comment-a',
  },
]

const User = objectType({
  name: 'User',
  definition(t) {
    t.implements('Node')
    t.id('id')
    t.string('name')
  },
})
const Post = objectType({
  name: 'Post',
  definition(t) {
    t.implements('Node')
    t.id('id')
    t.string('title')
  },
})
const Comment = objectType({
  name: 'Comment',
  definition(t) {
    t.id('id')
    t.string('text')
  },
})

const QueryAll = parse(
  `
  fragment nodeInfo on Node {
    __typename
    id
  }

  fragment userInfo on User {
    name
  }

  query QueryAll ($user1Id: ID! $user2Id: ID! $post1Id: ID! $post2Id: ID!) {
    user1: node(id: $user1Id) {
      ...nodeInfo
      ...userInfo
    }
    user2: node(id: $user2Id) {
      ...nodeInfo
      ...userInfo
    }
    post1: node(id: $post1Id) {
      ...nodeInfo
      ...userInfo
    }
    post2: node(id: $post2Id) {
      ...nodeInfo
      ...postInfo
    }

    nodes(ids: [$user1Id, $post1Id, $user2Id, $post2Id]) {
      ...nodeInfo
      ...userInfo
      ...postInfo
    }
  }`,
)
const QueryNode = parse(
  `
  fragment nodeInfo on Node {
    __typename
    id
  }

  query QueryNode ($nodeId: ID!) {
    node(id: $nodeId) {
      ...nodeInfo
    }
  }`,
)

const testSchema = (pluginConfig: RelayNodeInterfacePluginConfig, outputs = false) =>
  makeSchema({
    outputs,
    types: [
      User,
      Post,
      Comment,
      objectType({
        name: 'Query',
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        definition(_t) {},
      }),
    ],
    plugins: [relayNodeInterfacePlugin(pluginConfig)],
  })

const testIdFetcher: RelayNodeInterfacePluginConfig['idFetcher'] = ({ id, type }, _ctx, _info) => {
  if (type === 'User') return users.find((u) => u.id === id)
  if (type === 'Post') return posts.find((p) => p.id === id)
  if (type === 'Comment') return comments.find((p) => p.id === id)
  return null
}

// @ts-ignore
const testResolveType: RelayNodeInterfacePluginConfig['resolveType'] = (o) => {
  if (users.includes(o)) return 'User'
  if (posts.includes(o)) return 'Post'

  return null
}

const testPluginConfig = {
  idFetcher: testIdFetcher,
  resolveType: testResolveType,
}

describe('relayNodeInterfacePlugin', () => {
  describe('basic behavior', () => {
    it('should work correctly', () => {
      const schema = testSchema(testPluginConfig)
      expect(printType(schema.getType('Query')!)).toMatchSnapshot()
    })

    it('should resolve all fields correctly', async () => {
      const schema = testSchema(testPluginConfig)
      const nodes = await execute({
        schema,
        document: QueryAll,
        variableValues: {
          user1Id: toGlobalId('User', users[0].id),
          user2Id: toGlobalId('User', users[1].id),
          post1Id: toGlobalId('Post', posts[0].id),
          post2Id: toGlobalId('Post', posts[1].id),
        },
      })

      expect(nodes).toMatchSnapshot()
    })

    it('should resolve not found fields as null correctly', async () => {
      const schema = testSchema(testPluginConfig)
      const nodes = await execute({
        schema,
        document: QueryAll,
        variableValues: {
          user1Id: toGlobalId('User', users[0].id),
          user2Id: toGlobalId('User', users[1].id),
          post1Id: toGlobalId('Post', 'abcdef'),
          post2Id: toGlobalId('Post', posts[1].id),
        },
      })

      expect(nodes).toMatchSnapshot()
    })

    it('should throw error for invalid type', async () => {
      const schema = testSchema(testPluginConfig)
      const nodes = await execute({
        schema,
        document: QueryNode,
        variableValues: {
          nodeId: toGlobalId('Comment', comments[0].id),
        },
      })

      expect(nodes).toMatchSnapshot()
    })
  })
})
