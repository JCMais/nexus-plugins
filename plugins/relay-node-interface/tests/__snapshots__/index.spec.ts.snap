// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relayNodeInterfacePlugin basic behavior should resolve all fields correctly 1`] = `
Object {
  "data": Object {
    "nodes": Array [
      Object {
        "__typename": "User",
        "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
        "name": "user-a",
      },
      Object {
        "__typename": "Post",
        "id": "710c173a-c2af-11ea-86f1-83226be355fb",
      },
      Object {
        "__typename": "User",
        "id": "710c15fa-c2af-11ea-86f1-7f51ff98b69c",
        "name": "user-b",
      },
      Object {
        "__typename": "Post",
        "id": "a7a5c06e-c2d9-11ea-86f1-bfb838f3b7c6",
      },
    ],
    "post1": Object {
      "__typename": "Post",
      "id": "710c173a-c2af-11ea-86f1-83226be355fb",
    },
    "post2": Object {
      "__typename": "Post",
      "id": "a7a5c06e-c2d9-11ea-86f1-bfb838f3b7c6",
    },
    "user1": Object {
      "__typename": "User",
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "user2": Object {
      "__typename": "User",
      "id": "710c15fa-c2af-11ea-86f1-7f51ff98b69c",
      "name": "user-b",
    },
  },
}
`;

exports[`relayNodeInterfacePlugin basic behavior should resolve not found fields as null correctly 1`] = `
Object {
  "data": Object {
    "nodes": Array [
      Object {
        "__typename": "User",
        "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
        "name": "user-a",
      },
      null,
      Object {
        "__typename": "User",
        "id": "710c15fa-c2af-11ea-86f1-7f51ff98b69c",
        "name": "user-b",
      },
      Object {
        "__typename": "Post",
        "id": "a7a5c06e-c2d9-11ea-86f1-bfb838f3b7c6",
      },
    ],
    "post1": null,
    "post2": Object {
      "__typename": "Post",
      "id": "a7a5c06e-c2d9-11ea-86f1-bfb838f3b7c6",
    },
    "user1": Object {
      "__typename": "User",
      "id": "8a291df8-bbcf-11ea-9db0-ff6b23713451",
      "name": "user-a",
    },
    "user2": Object {
      "__typename": "User",
      "id": "710c15fa-c2af-11ea-86f1-7f51ff98b69c",
      "name": "user-b",
    },
  },
}
`;

exports[`relayNodeInterfacePlugin basic behavior should throw error for invalid type 1`] = `
Object {
  "data": Object {
    "node": null,
  },
  "errors": Array [
    [GraphQLError: Abstract type "Node" must resolve to an Object type at runtime for field "Query.node" with value { id: "abc", text: "comment-a" }, received "null". Either the "Node" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.],
  ],
}
`;

exports[`relayNodeInterfacePlugin basic behavior should work correctly 1`] = `
"type Query {
  \\"\\"\\"Fetches an object given its global ID\\"\\"\\"
  node(
    \\"\\"\\"The global ID of an object\\"\\"\\"
    id: ID!
  ): Node

  \\"\\"\\"Fetches objects given their global IDs\\"\\"\\"
  nodes(
    \\"\\"\\"The global IDs of objects\\"\\"\\"
    ids: [ID!]!
  ): [Node]!
}"
`;
